lib/commons/aria/roles.js

```diff
@@ -204,3 +202,248 @@ aria.implicitRole = function(node) {
 		ariaAttributes
 	).shift();
 };
+
+/**
+ * Get roles applied to a given node
+ * @method getRoleSegments
+ * @memberof axe.commons.aria
+ * @instance
+ * @param {HTMLElement} node HTMLElement
+ * @return {Array<String>} return an array of roles applied to the node, if no roles, return an empty array.
+ */
+aria.getRoleSegments = function(node) {
+	let roles = [];
+	if (node && node.hasAttribute('role')) {
+		roles = roles.concat(node.getAttribute('role').split(' '));
+	}
+	if (node && node.hasAttributeNS('http://www.idpf.org/2007/ops', 'type')) {
+		let rawEpubRoles = node.getAttributeNS(
+			'http://www.idpf.org/2007/ops',
+			'type'
+		);
+		let epubRoles = rawEpubRoles.split(' ');
+		roles = roles.concat(
+			epubRoles.map(role => {
+				return `doc-${role}`;
+			})
+		);
+	}
+	return roles;
+};
+
+/**
+ * Get roles applicable and or allowed for a given node/ HTMLElement
+ * @method getApplicableRolesForNode
+ * @instance
+ * @param {HTMLElement} node HTMLElement
+ * @return {Array<String>} return an array of roles applicable to the node, if no roles, return an empty array.
+ * @private
+ */
+function getApplicableRolesForNode(node) {
+	const tagName = node.tagName.toUpperCase();
+	const nodeAttrs = Array.from(node.attributes).map(a => a.name.toUpperCase());
+
+	const rolesTable = axe.commons.aria.lookupTable.role;
+	return Object.keys(axe.commons.aria.lookupTable.role).reduce((out, role) => {
```

Comment:
This reduce seems unnecessary. You're only ever doing `.includes()` with the result. You already know what role you are testing - all you have to do is pass the role in as an argument and do `lookupTable[role].allowedElements`.